ignore/src/hgignore.rs

```diff
@@ -0,0 +1,504 @@
+        lazy_static! {
+            static ref RE: Regex = Regex::new(
+                r"^syntax:\s*(regexp|glob)\s*$").unwrap();
+        };
+        let mut current_syntax = HgignoreSyntax::Regexp;
+        let path = path.as_ref();
+        let file = match File::open(path) {
+            Err(err) => return Some(Error::Io(err).with_path(path)),
+            Ok(file) => file,
+        };
+        let rdr = io::BufReader::new(file);
+        let mut errs = PartialErrorBuilder::default();
+        for (i, line) in rdr.lines().enumerate() {
+            let lineno = (i + 1) as u64;
+            let line = match line {
+                Ok(line) => line,
+                Err(err) => {
+                    errs.push(Error::Io(err).tagged(path, lineno));
+                    break;
+                }
+            };
+            match RE.captures(line.as_bytes()) {
+                None => {
+                    if let Err(err) = self.add_line(
+                        Some(path.to_path_buf()), &line, current_syntax
+                    ) {
+                        errs.push(err.tagged(path, lineno));
+                    };
+                },
+                Some(caps) => {
+                    current_syntax = match &caps[1] {
+                        b"glob" => HgignoreSyntax::Glob,
+                        _ => HgignoreSyntax::Regexp,
+                    }
+                },
+            };
+        }
+        errs.into_error_option()
+    }
+
+    /// doc
+    pub fn add_line(
+        &mut self,
+        from: Option<PathBuf>,
+        mut line: &str,
+        current_syntax: HgignoreSyntax
+    ) -> Result<&mut HgignoreBuilder, Error> {
+        if line.starts_with("#") {
+            return Ok(self);
+        }
+        if !line.ends_with("\\ ") {
+            line = line.trim_right();
+        }
+        if line.is_empty() {
+            return Ok(self);
+        }
+        match current_syntax {
+            HgignoreSyntax::Glob => self.add_glob_line(from, line),
+            HgignoreSyntax::Regexp => self.add_regex_line(from, line),
+        }
+    }
+
+    /// doc
+    pub fn add_regex_line(
+        &mut self,
+        from: Option<PathBuf>,
+        line: &str,
+    ) -> Result<&mut HgignoreBuilder, Error> {
+        let glob = Glob {
+            from: from,
+            original: line.to_string(),
+            actual: String::new(),
+            is_only_dir: false,
+        };
+        let parsed = try!(
+            GlobBuilder::new(&line.to_string())
+                // .literal_separator(literal_separator)
+                .from_regex(true)
+                .build()
+                .map_err(|err| {
+                    Error::Glob {
+                        glob: Some(glob.original.clone()),
+                        err: err.kind().to_string(),
+                    }
+                }));
+        self.builder.add(parsed);
+        self.globs.push(glob);
+        Ok(self)
+    }
+
+    /// doc
+    pub fn add_glob_line(
```

Comment:
There's a lot in common with this method and a similar method in `GitignoreBuilder`. Are the semantics really this identical? And if so, can we split the logic out into a function that both builders can use?